src/Serialization/Compact.php

```diff
@@ -41,13 +41,19 @@ public function __construct(
 
     /**
      * @param string $headersJson
+     *
      * @return HeaderParameter\ParameterInterface[]
+     * @throws \InvalidArgumentException
      */
     protected function parseHeaders($headersJson)
     {
         $parameters = [];
         $headers = json_decode($headersJson, true);
 
+        if (!is_array($headers) || empty($headers)) {
```

Comment:
It might be worth doing a quick [`json_last_error`](http://php.net/manual/en/function.json-last-error.php) test first. That would provide an explicit message explaining if the failure is due to unparseable json, rather than a generic catch-all, for instance if there are no headers provided but valid json. Same for the `parsePayload` method.